pyramid/csrf.py

```diff
@@ -0,0 +1,283 @@
+    is_same_domain,
+    strings_differ
+)
+
+
+@implementer(ICSRFStoragePolicy)
+class SessionCSRFStoragePolicy(object):
+    """ The default CSRF implementation, which mimics the behavior from older
+    versions of Pyramid. The ``new_csrf_token`` and ``get_csrf_token`` methods
+    are indirected to the underlying session implementation.
+
+    Note that using this CSRF implementation requires that
+    a :term:`session factory` is configured.
+
+    .. versionadded :: 1.8a1
+    """
+    def new_csrf_token(self, request):
+        """ Sets a new CSRF token into the session and returns it. """
+        return request.session.new_csrf_token()
+
+    def get_csrf_token(self, request):
+        """ Returns the currently active CSRF token from the session, generating
+        a new one if needed."""
+        return request.session.get_csrf_token()
+
+    def check_csrf_token(self, request, supplied_token):
+        """ Returns True if supplied_token is the same value as get_csrf_token
+        returns for this request. """
+        expected = self.get_csrf_token(request)
+        return not strings_differ(
+            bytes_(expected, 'ascii'),
+            bytes_(supplied_token, 'ascii'),
+        )
+
+@implementer(ICSRFStoragePolicy)
+class CookieCSRFStoragePolicy(object):
+    """ An alternative CSRF implementation that stores its information in
+    unauthenticated cookies, known as the 'Double Submit Cookie' method in the
+    OWASP CSRF guidelines. This gives some additional flexibility with regards
+    to scaling as the tokens can be generated and verified by a front-end
+    server.
+
+    .. versionadded :: 1.8a1
+    """
+
+    def __init__(self, cookie_name='csrf_token', secure=False, httponly=False,
+                 domain=None, max_age=None, path='/'):
+        serializer = _SimpleSerializer()
+        self.cookie_profile = CookieProfile(
+            cookie_name=cookie_name,
+            secure=secure,
+            max_age=max_age,
+            httponly=httponly,
+            path=path,
+            serializer=serializer
+        )
+        self.domain = domain
+
+    def new_csrf_token(self, request):
+        """ Sets a new CSRF token into the request and returns it. """
+        token = uuid.uuid4().hex
+        def set_cookie(request, response):
+            self.cookie_profile.set_cookies(
+                response,
+                token,
+            )
+        request.add_response_callback(set_cookie)
+        return token
+
+    def get_csrf_token(self, request):
+        """ Returns the currently active CSRF token by checking the cookies
+        sent with the current request."""
+        bound_cookies = self.cookie_profile.bind(request)
+        token = bound_cookies.get_value()
+        if not token:
+            token = self.new_csrf_token(request)
+        return token
+
+    def check_csrf_token(self, request, supplied_token):
+        """ Returns True if supplied_token is the same value as get_csrf_token
```

Comment:
"""Returns ``True`` if ``supplied_token`` is the same value as ``get_csrf_token``